src/zino/statemodels.py

```diff
@@ -363,3 +368,156 @@ class AlarmEvent(Event):
+        if self.match_type == "str":
+            return self._string_match(self.match_expression, device.name)
+        if self.match_type == "exact":
+            return self.match_expression == device.name
+        return False
+
+    def _get_or_create_events(self, state: "ZinoState") -> list[Event]:
+        """Creates/gets events that are affected by the given starting planned
+        maintenance
+        """
+        events = []
+        # all devices that the pm should affect
+        devices = [device for device in state.devices.devices.values() if self.matches_device(device)]
+        for device in devices:
+            reachability_event = state.events.get_or_create_event(device.name, None, ReachabilityEvent)
+            yellow_event = state.events.get_or_create_event(device.name, "yellow", AlarmEvent)
+            yellow_event.alarm_type = "yellow"
+            red_event = state.events.get_or_create_event(device.name, "red", AlarmEvent)
+            red_event.alarm_type = "red"
+            for event in (reachability_event, yellow_event, red_event):
+                events.append(event)
+        return events
+
+
+class PortStateMaintenance(PlannedMaintenance):
+    type: PmType = PmType.PORTSTATE
+
+    def matches_event(self, event: Event, state: "ZinoState") -> bool:
+        """Returns true if `event` will be affected by this planned maintenance"""
+        if event.type != "portstate":
+            return False
+        device = state.devices[event.router]
+        port = device.ports[event.ifindex]
+        return self.matches_portstate(device, port)
+
+    def matches_portstate(self, device: DeviceState, port: Port) -> bool:
+        """Returns true if PortstateEvents related to `port` on `device`
+        would be affected by this planned maintenance
+        """
+        if self.match_type == "regexp":
+            return self._regex_match(self.match_expression, port.ifdescr)
+        if self.match_type == "str":
+            return self._string_match(self.match_expression, port.ifdescr)
+        if self.match_type == "intf-regexp":
+            if self._regex_match(self.match_device, device.name):
+                return self._regex_match(self.match_expression, port.ifdescr)
+        return False
+
+    def _get_or_create_events(self, state: "ZinoState") -> list[Event]:
+        events = []
+        deviceports = self._get_matching_ports(state)
+        for device, port in deviceports:
+            event = state.events.get_or_create_event(device.name, port.ifindex, PortStateEvent)
+            event.ifindex = port.ifindex
+            events.append(event)
+        return events
+
+    def _get_matching_ports(self, state: "ZinoState") -> list[tuple[DeviceState, Port]]:
+        ports = []
+        for device in state.devices.devices.values():
+            for port in device.ports.values():
+                if self.matches_portstate(device, port):
+                    ports.append((device, port))
+        return ports
```

Comment:
It doesn't seem to need to keep an actual list of ports in memory here (also, multiple list appends are usually a lot slower than list comprehensions or generators). The `_get_matching_ports` function could be changed into a generator by simply `yield`ing each matched port

Reply:
fixed